### ISSUE TITLE
Recursive call to toc

### ISSUE THREAD

**Issue by cbeauchesne** [Python-Markdown/markdown#639]
In toc, extension, there is a security that checks toc presence in a header, to avoid endless loop.

But if the toc is hidden into anoter tag, it fails : 

    # *[toc]*

On this exemple, toc is  inside em tag, and thus missed by the check.

**Comment by waylan** [comment]
As it should. The toc is an `<ul>`, which is a block level element. Block level elements are not valid inside a span level element (such as italics). If we allowed that to work, we would be generating invalid HTML. Therefore, we don't check for the toc in span level elements, which gives us the added benefit of being slightly faster as there are less elements to check the contents of.

**Comment by cbeauchesne** [comment]
I agree that it's not a valid syntax and it should not build a toc inside a header 😊

The issue is it causes a MemoryError

**Comment by waylan** [comment]
Oh I just realized I misunderstood the report. The problem is that that input causes Markdown to crash. Yeah that is a bug.

**Comment by cbeauchesne** [comment]
Yes, and same effect for #640

**Comment by cbeauchesne** [comment]
After investigation, I think that this code would guarantee that no parent is forbidden, without any performance cost. 

It modifies `replace_marker()` method, and replace `iterparent()` method by `iter_allowed_parent()` method :

``` python
        def iter_allowed_parent(self, node):
            ''' Iterator wrapper to get parent and child all at once. '''

            for child in node:
                if not self.header_rgx.match(child.tag) and child.tag not in ['pre', 'code']:
                    yield node, child
                    for p, c in self.iter_allowed_parent(child):
                        yield p, c

        def replace_marker(self, root, elem):
            ''' Replace marker with elem. '''
            for (p, c) in self.iter_allowed_parent(root):
                text = ''.join(c.itertext()).strip()
                if not text:
                    continue

                # To keep the output from screwing up the
                # validation by putting a <div> inside of a <p>
                # we actually replace the <p> in its entirety.
                # We do not allow the marker inside a header as that
                # would causes an enless loop of placing a new TOC
                # inside previously generated TOC.
                if c.text and c.text.strip() == self.marker:
                    for i in range(len(p)):
                        if p[i] == c:
                            p[i] = elem
                            break
 ```

It solves #639 and #640.

Please apologize, I've not time to set up a functionnal dev environment, and can't run test case ...

**Comment by waylan** [comment]
If you could create a PR that would help get this fixed. Otherwise, someone will get to it when we have time.

**Comment by cbeauchesne** [comment]
> If you could create a PR

Tried : #641

**Comment by waylan** [comment]
Fixed in #641.
